game-core/src/main/java/games/strategy/engine/framework/ServerGame.java

```diff
@@ -307,6 +307,10 @@ public void startGame() {
           runStep(false);
         }
       }
+    } catch (final GameOverException e) {
```

Comment:
It looks like the `if` condition below could be modified to account for this, ie:
`if (e instanceof GameOverException || e.getCause() instanceof GameOverException) {`

Which makes it seem like the bug is that instead of `GameOverException` being the wrapped exception, it *is* the exception now. Considering we throw any other exception type, the entire catch block for `RuntimeException` and its nested if statement now looks unnecessary.

Reply:
Yeah I wasn't sure what to prefer at this point.
The whole exception thing is a complete mess.
I initially tried making GameOverException checked, but ultimately failed because at some points it's treated like an unchecked exception to terminate a game and at some other points in the code it gets treated as something unexpected that needs to be handled in a special way.
Combine that with the fact that at some point the exception is transmitted over the network and everything is completely unusable.

If you have the time you might want to have a look yourself, maybe you see something I missed

Reply:
I think you can keep it simple here, the RuntimeException block is now buggy because `e.getCause()` is no longer GameOverException.

Catching the right exception type is all that is needed now as any other exception is rethrown. This is an example of 'control-flow-by-exception-handling', just not clean.. Perhaps better would be to have a game stop listener, have that get executed and set a 'game-terminated' flag which would then exit this loop. That's a decent sized refactor rather than a bug fix. Pretty much I think you can remove the `catch(RuntimeException)` block and this should be good to go.

Reply:
> I initially tried making GameOverException checked, but ultimately failed because at some points it's treated like an unchecked exception to terminate a game and at some other points in the code it gets treated as something unexpected that needs to be handled in a special way.

:grinning:  Oh, TripleA! Not only do we have control-flow-by-exception-handling, but we have a violation of SRP too : )

It's a good call out how messy that is.
I suspect somewhere in there is the 'ghosting' bug, the system needs perhaps a good rework to be simpler and more pro-active to verify connectivity.

Reply:
In the PR that changed the exception types a bit there was a comment somewhere that raised the question if it would make sense to have GameOverException extend MessengerException.
After having to deal with this, I realised why just a couple hours ago: In local games this exception is just the "terminate game" exception.
In networked games this exception is used to somehow communicate disconnects that lead to the end of the game, that's why everything is so messed up.

Reply:
> I think you can keep it simple here, the RuntimeException block is now buggy because e.getCause() is no longer GameOverException.

Are you sure about that? I'm pretty sure I put that check there for a reason, but at the same time it could've been me being overly cautious

Reply:
I ran it through a debugger, the if statement fails because `e` is an instance of `GameOverException`

Reply:
Ok, but did you verify the case when you're hosting a networked game?
For local games, this was clearly my mistake, but I could imagine that for networked games the exception is wrapped at some point and rethrown when a player, or the host itself disconnects

Reply:
@DanVanAtta I'll get some sleep now.
Feel free to push the desired changes directly to this branch (you should be able to do so via the web client as well), if you want this merged ASAP.

Otherwise I'll commit the changes tomorrow based on your answer.

Reply:
@RoiEXLab this is your bug fix, please walk through a debugger in that case to verify.

Reply:
Good observations @DanVanAtta I wasn't able to hit this catch block either. Removed it.

Reply:
Thanks for verifying!